### PR TITLE
Partitioner add vg

### PR DESCRIPTION
**Pull request by joseivanlopez**
PBI: https://trello.com/c/CGjosZ0p/734-3-expert-partitioner-create-a-vg

### CODE REVIEW

File: src/lib/y2storage/disk_size.rb [hunk 1]
```diff
@@ -549,6 +549,25 @@ def zero?
...
+    #
+    # @example
+    #   x = DiskSize.KiB(4) #=> <DiskSize 4.00 KiB (4096)>
+    #   x.power_of_two?     #=> true
+    #
+    #   x = DiskSize.KB(4) #=> <DiskSize 3.91 KiB (4000)>
+    #   x.power_of_two?    #=> false
+    def power_of_two?
+      return false if zero? || unlimited?
+      size.to_s(2).count("1") == 1
```
Comment (ancorgs) [inline src/lib/y2storage/disk_size.rb]: I still haven't decided whether this is very neat or just hacky :wink: (the line between those two is always thin).

Why not something more generic and less surprising like this?
```ruby
def power_of?(exp)
  return false if unlimited?
  (Math.log(size, exp) % 1).zero?
end
```

File: src/lib/y2partitioner/widgets/md_add_button.rb [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+# find current contact information at www.suse.com.
+
+require "yast"
+require "cwm"
+require "y2partitioner/actions/add_md"
+
+module Y2Partitioner
+  module Widgets
+    # Button for openng a wizard to add a new MD array
+    class MdAddButton < CWM::PushButton
```
Comment (ancorgs) [inline src/lib/y2partitioner/widgets/md_add_button.rb]: Why wasn't this turned into a `DeviceButton`  like `LvmLvAddButton` and `PartitionAddButton`?

File: src/lib/y2partitioner/dialogs/lvm_vg.rb [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+        end
+
+        # @macro seeAbstractWidget
+        # Checks whether the given value is valid
+        #
+        # @note An error popup is shown when necessary.
+        #
+        # @return [Boolean]
+        def validate
+          focus
```
Comment (ancorgs) [inline src/lib/y2partitioner/dialogs/lvm_vg.rb]: Why?

File: src/lib/y2partitioner/dialogs/lvm_vg.rb [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+          Yast::UI.SetFocus(Id(widget_id))
+        end
+
+        # Checks whether some value was entered
+        #
+        # @return [Boolean]
+        def presence_validation
+          return true unless controller.empty_vg_name?
+
+          Yast::Popup.Error(
```
Comment (ancorgs) [inline src/lib/y2partitioner/dialogs/lvm_vg.rb]: NP: In general, having the detection of errors in the controller and the strings explaining them in the widgets feels wrong. We are forced to keep them in sync. We can live like that for the time being (deadlines are around the corner), but maybe some FIXMEs wouldn't hurt.

File: src/lib/y2partitioner/actions/controllers/lvm_vg.rb [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+        # @return [Y2Storage::DiskSize]
+        def vg_size
+          vg.size
+        end
+
+        # Applies given values (i.e., volume group name and extent size) to the
+        # volume group
+        def apply_values
+          vg.vg_name = vg_name
+          vg.extent_size = extent_size
```
Comment (ancorgs) [inline src/lib/y2partitioner/actions/controllers/lvm_vg.rb]: Assigning nil here to `extend_size` is correctly handled? Doesn't the library go nuts? Even more important, is that situation unit-tested?

File: test/y2partitioner/dialogs/lvm_vg_test.rb [hunk 6]
```diff
@@ -0,0 +1,542 @@
...
+      end
+
+      it "sets its value with the extent size stored in the controller" do
+        expect(subject).to receive(:value=).with("128 KiB")
+        subject.init
+      end
+    end
+
+    describe "#handle" do
+      before do
```
Comment (ancorgs) [inline test/y2partitioner/dialogs/lvm_vg_test.rb]: I would like to see a unit test for empty string and/or an invalid string here. Just to ensure it doesn't break today or in the future.

File: src/lib/y2partitioner/dialogs/lvm_vg.rb [hunk 7]
```diff
@@ -0,0 +1,267 @@
...
+        end
+
+        # Checks whether there is not another device with the entered name
+        #
+        # @return [Boolean]
+        def uniqueness_validation
+          return true unless controller.duplicated_vg_name?
+
+          Yast::Popup.Error(
+            format(
```
Comment (ancorgs) [inline src/lib/y2partitioner/dialogs/lvm_vg.rb]: Maybe a comment for translators here. My feeling is that they are used to the `%s` and `%1` notation, but not so much to the Ruby keyword placeholders one.

File: src/lib/y2partitioner/actions/controllers/lvm_vg.rb [hunk 8]
```diff
@@ -0,0 +1,298 @@
...
+        def illegal_vg_name?
+          vg_name.split(//).any? { |c| !ALLOWED_NAME_CHARS.include?(c) }
+        end
+
+        # Checks whether it exists another device with the given volume group name
+        #
+        # @return [Boolean] true if there is another device with the same name;
+        #   false otherwise
+        def duplicated_vg_name?
+          name = File.join("/dev", vg_name)
```
[the remark this answers is not in the document]
Comment (joseivanlopez) [reply src/lib/y2partitioner/actions/controllers/lvm_vg.rb]: Added comment to explain it.

File: src/lib/y2partitioner/dialogs/lvm_vg.rb (hunk 3 above)
Comment (joseivanlopez) [reply src/lib/y2partitioner/dialogs/lvm_vg.rb]: It is necessary to focus the field when an error happens. Commented in the code.

File: src/lib/y2partitioner/dialogs/lvm_vg.rb (hunk 4 above)
Comment (joseivanlopez) [reply src/lib/y2partitioner/dialogs/lvm_vg.rb]: I agree, so the messages generation has been moved to the controller.

File: src/lib/y2partitioner/dialogs/lvm_vg.rb (hunk 7 above)
Comment (joseivanlopez) [reply src/lib/y2partitioner/dialogs/lvm_vg.rb]: Added.

File: src/lib/y2partitioner/widgets/md_add_button.rb (hunk 2 above)
Comment (joseivanlopez) [reply src/lib/y2partitioner/widgets/md_add_button.rb]: As commented, `LvmVgAddButton` and `MdAddButton` do not work over a specific device. The class `DeviceButton` is meant to act over a device.

File: src/lib/y2storage/disk_size.rb (hunk 1 above)
Comment (joseivanlopez) [reply src/lib/y2storage/disk_size.rb]: Totally agree (but you cannot say that the hack was quite cute)

File: test/y2partitioner/dialogs/lvm_vg_test.rb (hunk 6 above)
Comment (joseivanlopez) [reply test/y2partitioner/dialogs/lvm_vg_test.rb]: Added.

File: src/lib/y2partitioner/actions/controllers/lvm_vg.rb (hunk 5 above)
Comment (joseivanlopez) [reply src/lib/y2partitioner/actions/controllers/lvm_vg.rb]: This method is only called when all validations are passed, so in practice it is not possible to assign a nil value.

File: src/lib/y2storage/disk_size.rb (hunk 1 above)
Comment (ancorgs) [reply src/lib/y2storage/disk_size.rb]: Yes. It was cool. Unnecessary but cool :-)

File: src/lib/y2partitioner/actions/controllers/lvm_vg.rb [hunk 9]
```diff
@@ -0,0 +1,364 @@
...
+          # generate the volume group name has been duplicated here due to its simplicity.
+          name = File.join("/dev", vg_name)
+          Y2Storage::BlkDevice.all(working_graph).map(&:name).include?(name)
+        end
+
+        # Error message to show when the given volume group name is duplicated
+        #
+        # @return [String]
+        def duplicated_vg_name_message
+          # TRANSLATORS: vg_name is replaced by the name of the volume group
```
Comment (ancorgs) [inline src/lib/y2partitioner/actions/controllers/lvm_vg.rb]: NP: I would say: `%{vg_name} is replaced...` The whole point is that translators may not be familiar with the `%{whatever}` syntax.

**Pull request merged**